thibaultyou/prompt-library
# Add Optional/Required Metadata for Prompt Input Variables and Custom Instruction Extension

### 🎯 Objective
This pull request aims to enhance our metadata generation and view update processes by adding information about whether prompt input variables are optional or required, and introducing custom instruction extension capabilities for all prompts. These changes will improve the clarity, usability, and flexibility of our prompt library.

### 🏗 Architectural Impact
This change primarily affects the metadata structure and generation process:

- Metadata generation (`core/generate_metadata.ts`)
- View update process (`core/update_views.ts`)
- Prompt analysis agent (`system_prompts/prompt_analysis_agent/prompt.md`)
- View template (`templates/sub_readme.md`)
- Metadata type definition (`types/metadata.ts`)

### 📊 Detailed Changes

#### 🔧 Core Functionality

1. Update Metadata Interface:
   ```typescript
   // types/metadata.ts
   export interface Metadata {
     // ... existing fields
     variables: Array<{
       name: string;
       role: string;
       optional_for_user: boolean; // New field
     }>;
   }
   ```

2. Modify Prompt Analysis Agent:
   Update `system_prompts/prompt_analysis_agent/prompt.md` to include logic for determining if variables are optional or required.

3. Update Metadata Generation:
   Modify `core/generate_metadata.ts` to process the new optional/required information.

4. Enhance View Template:
   Update `templates/sub_readme.md` to display "(Optional)" next to variable names where `optional_for_user` is true.

5. Implement Custom Instruction Extension:
   Add capability to extend prompts with custom instructions through automation and prompt generation.

### 🧪 Testing Strategy
- Update existing tests to account for the new `optional_for_user` field
- Add new tests to verify correct identification of optional and required variables
- Test custom instruction extension functionality

### 📚 Documentation Updates
- Update README to explain the new optional/required variable feature
- Document the custom instruction extension capability

### 🔮 Future-Proofing
- The new metadata structure allows for easier addition of variable-specific attributes in the future
- Custom instruction extension provides a foundation for more dynamic prompt customization

### 🔗 Related Issues
Closes #26
Closes #34

This pull request addresses the core requirements outlined in the issues without introducing speculative features or unmentioned enhancements. It focuses on implementing the optional/required flags for input variables and adding custom instruction extension capabilities, which will significantly improve the usability and flexibility of our prompt library.
